core/src/banking_simulation.rs

```diff
@@ -0,0 +1,926 @@
+    freeze_time_by_slot: FreezeTimeBySlot,
+    base_event_time: SystemTime,
+    base_simulation_time: SystemTime,
+}
+
+impl SimulatorLoopLogger {
+    fn new(
+        simulated_leader: Pubkey,
+        base_event_time: SystemTime,
+        base_simulation_time: SystemTime,
+        freeze_time_by_slot: FreezeTimeBySlot,
+    ) -> Self {
+        Self {
+            simulated_leader,
+            base_event_time,
+            base_simulation_time,
+            freeze_time_by_slot,
+        }
+    }
+
+    fn bank_costs(bank: &Bank) -> (u64, u64) {
+        bank.read_cost_tracker()
+            .map(|t| (t.block_cost(), t.vote_cost()))
+            .unwrap()
+    }
+
+    fn log_frozen_bank_cost(&self, bank: &Bank) {
+        info!(
+            "bank cost: slot: {} {:?} (frozen)",
+            bank.slot(),
+            Self::bank_costs(bank),
+        );
+    }
+
+    fn log_ongoing_bank_cost(&self, bank: &Bank) {
+        debug!(
+            "bank cost: slot: {} {:?} (ongoing)",
+            bank.slot(),
+            Self::bank_costs(bank),
+        );
+    }
+
+    fn log_jitter(&self, bank: &Bank) {
+        let old_slot = bank.slot();
+        if let Some(event_time) = self.freeze_time_by_slot.get(&old_slot) {
+            if log_enabled!(log::Level::Info) {
+                let current_simulation_time = SystemTime::now();
+                let elapsed_simulation_time = current_simulation_time
+                    .duration_since(self.base_simulation_time)
+                    .unwrap();
+                let elapsed_event_time = event_time.duration_since(self.base_event_time).unwrap();
+                info!(
+                    "jitter(parent_slot: {}): {}{:?} (sim: {:?} event: {:?})",
+                    old_slot,
+                    if elapsed_simulation_time > elapsed_event_time {
+                        "+"
+                    } else {
+                        "-"
+                    },
+                    if elapsed_simulation_time > elapsed_event_time {
+                        elapsed_simulation_time - elapsed_event_time
+                    } else {
+                        elapsed_event_time - elapsed_simulation_time
+                    },
+                    elapsed_simulation_time,
+                    elapsed_event_time,
+                );
+            }
+        }
+    }
+
+    fn on_new_leader(&self, bank: &Bank, new_slot: Slot, new_leader: Pubkey) {
+        self.log_frozen_bank_cost(bank);
+        info!(
+            "{} isn't leader anymore at slot {}; new leader: {}",
+            self.simulated_leader, new_slot, new_leader
+        );
+    }
+}
+
+struct SenderLoopLogger<'a> {
+    non_vote_sender: &'a TracedSender,
+    tpu_vote_sender: &'a TracedSender,
+    gossip_vote_sender: &'a TracedSender,
+    last_log_duration: Duration,
+    last_tx_count: usize,
+    last_non_vote_count: usize,
+    last_tpu_vote_tx_count: usize,
+    last_gossip_vote_tx_count: usize,
+    non_vote_count: usize,
```

Comment:
are the batch counts useful?
If they are, please add some sort of indication these are batch counts. It's very confusing what the difference is currently.

Reply:
yeah, these are useful to some extent to gauge how we're batching txes: d949c1fc04ce0aa1a2af59bece25919a9bf66774